Syncing with v4.2 (#1373)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-cookbook-defaults.adoc

endif::[]
:keywords: studio, anypoint, transform, transformer, format, aggregate, rename, split, filter convert, xml, json, csv, pojo, java object, metadata, dataweave, data weave, datamapper, dwl, dfl, dw, output structure, input structure, map, mapping, examples

> 

Reviewer: The topic set title "Set a Default Value" implies a Task, because it begins with an imperative: do this (Set).
That makes me, as a reader, expect steps, which are not forthcoming.
Going forward, we are going to separate our narratives into distinct Task, Concept, and Reference topics and follow the titling practices defined in the style ref, using noun-based titles for Concept and Reference topics and imperative for Tasks.

Reviewer: Better title: Methods for Setting Default Values

> You can define default values to use in case any value you are querying is not present.

Reviewer: ```suggestion
Use one of the following methods to set default values so that queries do not fail in the absence of a parameter value:
```
OR, upon reading further, perhaps this is more accurate?

```suggestion
Use one of the following methods to set default values so that queries do not fail in the absence of a payload value or when the value is `null`:
```

> Set default values by any of the following methods:

Reviewer: See above.

Also, beware of defaulting to the practice of separating sentences that might belong together as a paragraph.

Reviewer: Because the list items in Lines 11-13 are methods and not user actions, they are fragments and should not end with periods.

> == Using the Key Word default

Reviewer: Again, the gerund "using" might imply that the audience can expect a task, with steps (although we would use our style of using the imperative, Use the Keyword default).
In this case, the title should tell those skimming the titles that this is a method, not a user action, so...:
```suggestion
== Example of Using the Keyword default to Set a Default Value
```

> Consider a Mule application that expects a JSON input with fields `id` and `name` to do a transformation. +

Reviewer 2: I aski
```suggestion
Consider a Mule application that expects a JSON input with fields `id` and `name` to do a transformation. 
```

Reviewer 2: I asked Kim about that `+`, and looks like it goes against our normal sentence style because it prevents normal wrapping.

Reviewer: Yes, GLOBALLY avoid using a space followed by "+" to force a line break, because it actually creates a paragraph break (which you want to create by pressing `return` twice, rather than using the Asciidoctor syntax +).
Carefully consider where you break for a new paragraph. Such a break should signal the transition to a new focus.

Reviewer: GLOBALLY avoid using "do" when you really mean "perform":
```suggestion
Consider a Mule application that expects a JSON input that contains the fields `id` and `name` to perform a DataWeave transformation, but the value for these fields is `null`. Configuring default values for these fields enables the transformation to occur:
```
Is this technically accurate? I got confused by us saying that these fields have the value `null` but we configure defaults that are actually also null: 0000 and Undefined. Does providing these values prevent an error or failure that would be caused by the value `null`?

Reviewer: When I read below, we don't actually mention a value of `null` but rather "an empty message". Which is correct, if they are the same thing?
Also, why do we give an example of fields that actually have values if this text is about providing defaults when they do not? If you want to show the example with fields, I suggest starting this topic with "Consider a Mule application that expects JSON input containing values for the fields `id` and `name`:", and then show the "normal" result. THEN mention the issue of what happens if there are NO values for these fields, and how providing defaults solves this.

Reviewer: I would not make these bullet list items, but a running narrative from stating the normal situation, with values, and then describing the less normal situation and its solution by providing defaults.

You can configure default values for these fields in case the payload value is `null`. For example:

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
{
  "userId": payload.id default "0000",
  "userName": payload.name default "Undefined"
}

> ----

Reviewer: See suggestions above.

* If the Mule application receives the following JSON input:
+
[source,json, linenums]
----
{
  "id": "123",
  "name": "Max the Mule"
}
----
+

> The DataWeave transformation output is:

Reviewer: See suggestions above.

Also, GLOBALLY do not break an implied "if...then" sentence around an example (or break ANY sentence around ANY kind of graphic display. In this case, "*If* the Mule application receives the following JSON input:" <example> "[implied *then*]The DataWeave transformation output is:" <example>

+
[source,JSON,linenums]
----
{
  "userId": "123",
  "userName": "Max the Mule"
}
----

> * If the Mule application receives an empty JSON message, the DataWeave transformation output is:

Reviewer: See suggestions above and ditto breaking the sentence around the examples.

+
[source,JSON,linenums]
----
{
  "userId": "0000",
  "userName": "Undefined"
}
----

> == Setting a Default Within if-else and else-if Statements

Reviewer: Ditto retitling: Example of Setting....

xref:dataweave-flow-control.adoc#control_flow_else_if[else-if]

> statements to specify a default value.

Reviewer: "Another method for providing a default value is using xref:dataweave-flow-control.adoc#control_flow_if_else[if-else] and xref:dataweave-flow-control.adoc#control_flow_else_if[else-if]. 

Under typical circumstances, providing a value ...results in expected output:
<example>

But if no value is specified, having a default configured results in usable output rather than <what>:"

> The following example sets `US` as the default value for `userLocation` if it is not present in the JSON input message:

Reviewer: Below, DITTO previous suggestion of presenting the behavior of a provided value first, and then presenting the default solution for a value not provided.

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
if (payload.location != null) {
	"userLocation" : payload.location
} else {
	"userLocation" : "United States"
}
----

> * If the Mule Application receives the following JSON input:

Reviewer: Also DITTO re. not breaking sentences around examples.

Reviewer: ```suggestion
* If the Mule application receives the following JSON input:
```

+
[source,json, linenums]
----
{
  "location": "Argentina"
}
----
+
The DataWeave transformation output is:
+
[source,JSON,linenums]
----
{
  "userLocation": "Argentina"
}
----

> * If the Mule Application receives an empty JSON message, the DataWeave transformation output is:

Reviewer: ```suggestion
* If the Mule application receives an empty JSON message, the DataWeave transformation output is:
```

+
[source,JSON,linenums]
----
{
  "userLocation": "United States"
}
----

> == Setting a Default When Matching Patterns

Reviewer: Please apply suggestions on previous two subtopics to this one as well:
title
method description
typical behavior
atypical values missing and method solution

Reviewer: Remainder copyedited only
